### PR TITLE
Cli describe

### PR DESCRIPTION
**Pull request by neilbacon**
Questions in # TODO: comments.

Works: python{,3} -m clkhash.cli describe fake-clk.json

Works: python3 -m unittest tests/test_bloomfilter.py
but with python (2):
  ...
  File "/usr/lib/python2.7/unittest/loader.py", line 91, in loadTestsFromName
    module = __import__('.'.join(parts_copy))
ImportError: Import by filename is not supported.

Type checker not yet run.

All improvements/comments welcome.

### CODE REVIEW

File: clkhash/bloomfilter.py [hunk 1]
```diff
@@ -352,3 +352,13 @@ def serialize_bitarray(ba):
 
     """
     return base64.b64encode(ba.tobytes()).decode('utf8')
+
+# TODO: I think this must already exist in anonlink - maybe that implelemtation should be moved here
```
Comment (hardbyte) [inline clkhash/bloomfilter.py]: Good point - probably into a `serialization.py` module though

File: setup.py [hunk 2]
```diff
@@ -13,7 +13,8 @@
         "jsonschema>=2.6",
         "requests>=2.18",
         "tqdm>=4.24",
-        "typing>=3.6; python_version < '3.5'"  # Backport from Py3.5
+        "typing>=3.6; python_version < '3.5'",  # Backport from Py3.5
+	"bashplotlib>=0.6.5"
```
Comment (hardbyte) [inline setup.py]: Think a tab snuck in here

File: clkhash/describe.py [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+import json
+from clkhash.backports import raise_from
+from bashplotlib.histogram import plot_hist
+from clkhash.bloomfilter import deserialize_bitarray
+
+class DescribeError(Exception):
+    """ The user provided CLK JSON is invalid.
+    """
+
+def plot(input):
```
Comment (hardbyte) [inline clkhash/describe.py]: I like this - much cleaner having the command implemented in its own module.

File: clkhash/cli.py [hunk 4]
```diff
@@ -270,6 +270,12 @@ def benchmark():
     bench.compute_hash_speed(10000)
 
 
+@cli.command('describe', short_help='show distribution of clk popcounts')
+@click.argument('input', type=click.File('r'))
+def describe(input):
```
Comment (hardbyte) [inline clkhash/cli.py]: Don't use `input` as it shadows the builtin `input` function.

Consider adding a `--no-plot` flag

Comment (neilbacon) [reply clkhash/cli.py]: Changed input to clk_json in describe.py.

In cli.py `input` is also used in `hash` and `upload` commands (I copied). Changing it requires a matching change to `@click.argument('input', type=click.File('r'))` which changes the `cli cmd --help` output. Do we want to do that? If so how about:
hash: rename input -> pii_csv
upload: rename input -> clk_json
describe: rename input -> clk_json

Comment (hardbyte) [reply clkhash/cli.py]: Sure let's not bother with reimplementing the summary table.

Your proposed parameter names sound good ' up to you if you'd like to make the change for the other commands.

File: clkhash/describe.py (hunk 3 above)
Comment (hardbyte) [reply clkhash/describe.py]: Yeah I can't tell what the trailing digit is, I guess the numbers are better than nothing!

Regarding testing I'd just like a simple smoketest - something which uses the output file from the hash command and tries to `describe` it.

File: clkhash/bloomfilter.py (hunk 1 above)
Comment (neilbacon) [reply clkhash/bloomfilter.py]: OK I'll move serialization to serialization.py.

I've noticed inconsistent presence of: 
#!/usr/bin/env python3
or #!/usr/bin/env python3.6
or nothing at top of files.
Should that be removed?

File: clkhash/cli.py (hunk 4 above)
Comment (hardbyte) [reply clkhash/cli.py]: Yikes! I'm glad you're catching all of these. Yes those suggestions all sound fine. 

Unfortunately some of these outputs *could* change soon with the addition of supporting a binary clk file. #176

File: clkhash/bloomfilter.py (hunk 1 above)
Comment (hardbyte) [reply clkhash/bloomfilter.py]: Sure, I don't feel strongly at all about those

File: clkhash/describe.py (hunk 3 above)
Comment (neilbacon) [reply clkhash/describe.py]: OK, test added, discovered StringIO and setting sys.stdout in the process.

File: clkhash/cli.py (hunk 4 above)
Comment (neilbacon) [reply clkhash/cli.py]: these changes now in cli.py (added pii_schema.json to example cmd line too)

File: clkhash/serialization.py [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+
+
+def serialize_bitarray(ba):
+    # type: (bitarray) -> str
+    """Serialize a bitarray (bloomfilter)
+
+    """
+    return base64.b64encode(ba.tobytes()).decode('utf8')
+
+# TODO: I think this must already exist in anonlink - maybe that implelemtation should be moved here
```
Comment (hardbyte) [inline clkhash/serialization.py]: Remove this before note before merging

Comment (neilbacon) [reply clkhash/serialization.py]: The new code has a test that serialized length (old code) is correct which fails for python 2, still deserializes correctly, but the serialized form isn't compatible between python 2 & 3. Haven't figured this out yet.

**Pull request merged**